### PR TITLE
Avoid expensive reshape for odd number of rows in conv2D

### PR DESCRIPTION
**Pull request by astojilj**
This PR addresses tensorflow/tfjs#949; conv2D is the main source of expensive reshapes in e.g. DeepLabV3 + MobileNetV2 model: before the patch, with environmental flags 'WEBGL_PACK:true,WEBGL_PACK_DEPTHWISECONV:false', there are 70 expensive reshapes per inference - with the patch, there are only 4.

Performance is measured using [modified conv2d benchmark](https://gist.github.com/astojilj/4ce6b26807b9f9636dbe81c7efba4166) in such way that it uses conv2d by matmul implementation: 

##### master branch, unpacked; [(tf.ENV.set('WEBGL_PACK', false)](https://gist.github.com/astojilj/4ce6b26807b9f9636dbe81c7efba4166#file-packedconv2dbenchmark-patch-L25);
```
    "N=227 {\"inDepth\":16,\"outDepth\":32,\"filterSize\":1,\"stride\":1,\"pad\":\"same\"}": {
      "averageTimeMs": 103.76119999957155
    },
    "N=513 {\"inDepth\":16,\"outDepth\":32,\"filterSize\":1,\"stride\":1,\"pad\":\"same\"}": {
      "averageTimeMs": 616.3369999990391
    }
```
##### master branch, packed ([tf.ENV.set('WEBGL_PACK', true)](https://gist.github.com/astojilj/4ce6b26807b9f9636dbe81c7efba4166#file-packedconv2dbenchmark-patch-L25)) 
```
    "N=227 {\"inDepth\":16,\"outDepth\":32,\"filterSize\":1,\"stride\":1,\"pad\":\"same\"}": {
      "averageTimeMs": 49.42674999969313
    },
    "N=513 {\"inDepth\":16,\"outDepth\":32,\"filterSize\":1,\"stride\":1,\"pad\":\"same\"}": {
      "averageTimeMs": 216.8936500007112
    }
```

##### astojilj:conv-avoid-expensive-reshape branch, packed ([tf.ENV.set('WEBGL_PACK', true)](https://gist.github.com/astojilj/4ce6b26807b9f9636dbe81c7efba4166#file-packedconv2dbenchmark-patch-L25)) 
```
    "N=227 {\"inDepth\":16,\"outDepth\":32,\"filterSize\":1,\"stride\":1,\"pad\":\"same\"}": {
      "averageTimeMs": 32.91375000029802
    },
    "N=513 {\"inDepth\":16,\"outDepth\":32,\"filterSize\":1,\"stride\":1,\"pad\":\"same\"}": {
      "averageTimeMs": 130.68550000083633
    }
```

PERF

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1495)

### CODE REVIEW

**Comment by astojilj** [comment]
@annxingyuan, PTAL

**Comment by astojilj** [comment]
@annxingyuan FYI, benchmark added.

File: src/kernels/backend_webgl.ts [hunk 1]
```diff
@@ -1630,6 +1630,55 @@ export class MathBackendWebGL implements KernelBackend {
...
+      return this.batchMatMul($x, $filter, false, false).reshape<Rank.R4>(
+          convInfo.outShape);
+    }
+    
+    const xshape = x.shape;
+    const xTexData = this.texData.get(x.dataId);
+    // We avoid expensive packed reshape by padding row count to next even
+    // number. When number of rows in batch, x.shape[2]), in 2x2 packed texture
+    // layout is odd, the result of packed batchMatMul is the same as it is for
+    // next, even, number of rows in the batch - x.shape[2] + 1.
```
Comment (annxingyuan) [inline src/kernels/backend_webgl.ts]: This comment is a little bit confusing. Are you saying that if we add an empty row to x and apply pointwise convolution, we can recover the correct result by dropping the last row of the output? Could you rephrase this comment using an example with concrete shapes for `x` and `filter`? ASCII art never hurts, but it's fine if you can just explain this a little more clearly :)

File: src/kernels/backend_webgl.ts [hunk 2]
```diff
@@ -1630,6 +1630,55 @@ export class MathBackendWebGL implements KernelBackend {
...
+    const packedXRowPad = xshape[2] % 2 === 1 && xTexData.isPacked ? 1 : 0;
+
+    const $x = Tensor.make(
```
Comment (annxingyuan) [inline src/kernels/backend_webgl.ts]: Please rename this to `xReshaped`

File: src/kernels/backend_webgl.ts [hunk 3]
```diff
@@ -1630,6 +1630,55 @@ export class MathBackendWebGL implements KernelBackend {
...
+        [1, xshape[0] * xshape[1] * (xshape[2] + packedXRowPad),
+         convInfo.inChannels], {dataId: x.dataId}, x.dtype) as Tensor3D;
+    if (packedXRowPad) {
+        xTexData.shape[xTexData.shape.length - 2]++;
+        util.assert(webgl_util.isReshapeFree(xTexData.shape, $x.shape),
+            `packed reshape from ${xTexData.shape} to ${$x.shape} isn't free`,);
+    }
+    const $filter = this.reshape(filter,
```
Comment (annxingyuan) [inline src/kernels/backend_webgl.ts]: Please rename this to `filterReshaped`

File: src/kernels/backend_webgl.ts [hunk 4]
```diff
@@ -1630,6 +1630,55 @@ export class MathBackendWebGL implements KernelBackend {
...
+        [1, convInfo.inChannels, convInfo.outChannels]) as Tensor3D;
+    
+    const xf = this.batchMatMul($x, $filter, false, false);
```
Comment (annxingyuan) [inline src/kernels/backend_webgl.ts]: What does `xf` stand for? Could you rename this to something a little more descriptive, like `pointwiseConv`?

File: src/kernels/backend_webgl.ts [hunk 5]
```diff
@@ -1630,6 +1630,55 @@ export class MathBackendWebGL implements KernelBackend {
...
+    const xfTexData = this.texData.get(xf.dataId);
+    util.assert(xfTexData.isPacked, 'expected packed result of batchMatMul');
```
Comment (annxingyuan) [inline src/kernels/backend_webgl.ts]: Why this assertion? Under what circumstances does this fail?

File: src/ops/conv2d_test.ts [hunk 6]
```diff
@@ -74,6 +74,33 @@ describeWithFlags('conv2d', ALL_ENVS, () => {
     expectArraysClose(result, [2, 4, 6, 8]);
   });
 
+  it('packed x=[2,1,2] f=[1,1,2,2] s=1 d=1 p=0', () => {
```
Comment (annxingyuan) [inline src/ops/conv2d_test.ts]: Could you move this test into a `describeWithFlags` block that applies only to `WEBGL_ENVS`? I think this is currently under `ALL_ENVS`

File: src/ops/conv2d_test.ts [hunk 7]
```diff
@@ -74,6 +74,33 @@ describeWithFlags('conv2d', ALL_ENVS, () => {
...
+
+    const webglLazilyUnpackFlagSaved = tf.ENV.get('WEBGL_LAZILY_UNPACK');
+    tf.ENV.set('WEBGL_LAZILY_UNPACK', true);
+    const webglLazilyPackBinaryOperationsFlagSaved =
+        tf.ENV.get('WEBGL_PACK_BINARY_OPERATIONS');
+    tf.ENV.set('WEBGL_PACK_BINARY_OPERATIONS', true);
+    
+    // |result| is packed for WebGL backends.
+    const result = tf.conv2d(x, w, stride, pad);
+    const result1 = tf.conv2d(result, w, stride, pad);
```
Comment (annxingyuan) [inline src/ops/conv2d_test.ts]: Can you help me understand the role of this test? Why does applying convolution twice uniquely test your new kernel? There should be other tests in our conv2d suite that hit your kernel, right? Please include your explanation in a comment for this test.

Comment (astojilj) [reply src/ops/conv2d_test.ts]: Changed the comment:

```
    // First conv2D tests conv2D with non-packed input |x|, and the second uses
    // packed input |result|.
    const result = tf.conv2d(x, w, stride, pad);
    const result1 = tf.conv2d(result, w, stride, pad);

```

MathBackendWebGL.conv2dByMatMul optimizes the case when input is packed.
All other cases hitting the test have non-packed input.
That said, I could replace the first "`const result = tf.conv2d(x, w, stride, pad);`" by any other operation producing packed output.

File: src/kernels/backend_webgl.ts (hunk 1 above)
Comment (astojilj) [reply src/kernels/backend_webgl.ts]: Didn't use ascii art but [rephrased the comment](https://github.com/tensorflow/tfjs-core/pull/1495/files#diff-9d35adb9ccc3d36d43672a2b9ba683fbR1650).
Note that the code is simplified with the latest commit, making it easier to read.

File: src/kernels/backend_webgl.ts [hunk 8]
```diff
@@ -1630,6 +1630,55 @@ export class MathBackendWebGL implements KernelBackend {
...
+        !ENV.get('WEBGL_PACK_BINARY_OPERATIONS')) {
+      const $x = this.reshape(x, [1, x.shape[0] * x.shape[1] * x.shape[2],
+          convInfo.inChannels]) as Tensor3D;
+      const $filter = this.reshape(filter,
+          [1, convInfo.inChannels, convInfo.outChannels]) as Tensor3D;
+      return this.batchMatMul($x, $filter, false, false).reshape<Rank.R4>(
+          convInfo.outShape);
+    }
+    
+    const xshape = x.shape;
```
[the remark this answers is not in the document]
Comment (astojilj) [reply src/kernels/backend_webgl.ts]: Thanks

**Comment by annxingyuan** [comment]
Thanks @astojilj !

**Comment by astojilj** [comment]
@annxingyuan ,
There is a serious, reproducible issue with this patch ([using DeepLabV3+ MobileNetV2 model](https://github.com/01org/depth-camera-web-demo/tree/master/nn/using-deeplab)) - wrong calculation of MathBackendWebGL.numBytesInGPU and [triggering paging](https://github.com/tensorflow/tfjs-core/blob/cdb8c4ab1a55bb6c43eb469a8ba34439b031157a/src/kernels/backend_webgl.ts#L2207).

const xf = this.batchMatMul($x, $filter, false, false) - [acquires texture](https://github.com/tensorflow/tfjs-core/blob/cdb8c4ab1a55bb6c43eb469a8ba34439b031157a/src/kernels/backend_webgl.ts#L2368) of logical shape with padded row count in texData.shape and later, after the texData.shape is decreased, [releaseTexture](https://github.com/tensorflow/tfjs-core/blob/cdb8c4ab1a55bb6c43eb469a8ba34439b031157a/src/kern
els/backend_webgl.ts#L2357) subtract lower number from numBytesInGPU... in time, leading to increase of numBytesInGPU and triggering the paging.

I have prepared a fix for it, and gpu allocation in general, in PR #1559.

Please, revert this patch, until the issue is resolved, or check the fix in PR #1559.
Thanks.

@dsmilkov,
FYI, of you plan to release, this should be a blocker.
